crates/noirc_frontend/src/hir/def_collector/dc_mod.rs

```diff
@@ -54,15 +57,113 @@ pub fn collect_defs(
 
     collector.collect_globals(context, ast.globals, errors);
 
+    collector.collect_traits(ast.traits, crate_id, errors);
+
     collector.collect_structs(ast.types, crate_id, errors);
 
     collector.collect_type_aliases(context, ast.type_aliases, errors);
 
     collector.collect_functions(context, ast.functions, errors);
 
+    collector.collect_trait_impls(context, ast.trait_impls, errors);
+
     collector.collect_impls(context, ast.impls);
 }
 
+fn check_trait_method_implementation_parameters(
+    expected_parameters: &Vec<(Ident, UnresolvedType)>,
+    impl_method: &NoirFunction,
+    trait_name: &str,
+) -> Result<(), DefCollectorErrorKind> {
+    let expected_num_parameters = expected_parameters.len();
+    let actual_num_parameters = impl_method.def.parameters.len();
+    if actual_num_parameters != expected_num_parameters {
+        return Err(DefCollectorErrorKind::MismatchTraitImplementationNumParameters {
+            actual_num_parameters,
+            expected_num_parameters,
+            trait_name: trait_name.to_owned(),
+            impl_ident: impl_method.name_ident().clone(),
+        });
+    }
+    for (count, (parameter, typ, _abi_vis)) in impl_method.def.parameters.iter().enumerate() {
+        let (_expected_name, expected_type) = &expected_parameters[count];
+        if typ != expected_type {
+            return Err(DefCollectorErrorKind::MismatchTraitImlementationParameter {
+                trait_name: trait_name.to_owned(),
+                expected_type: expected_type.clone(),
+                impl_method: impl_method.name().to_string(),
+                parameter: parameter.name_ident().clone(),
+            });
+        }
+    }
+    Ok(())
+}
+
+fn check_trait_method_implementation_return_type(
+    expected_return_type: &FunctionReturnType,
+    impl_method: &NoirFunction,
+    trait_name: &str,
+) -> Result<(), DefCollectorErrorKind> {
+    match expected_return_type {
+        FunctionReturnType::Default(_span) => match &impl_method.def.return_type {
+            FunctionReturnType::Default(_span) => Ok(()),
+            FunctionReturnType::Ty(_found_type, _span) => {
+                Err(DefCollectorErrorKind::MismatchTraitImplementationReturnType {
+                    trait_name: trait_name.to_owned(),
+                    impl_ident: impl_method.name_ident().clone(),
+                })
+            }
```

Comment:
These could still be equal if the impl method's return type is unit. Checking `impl_method.return_type()` may be simpler here since you have to check the type anyway

Reply:
Another alternative I suppose is to defer this until type checking. It is a bit odd to check types before then anyway.